### PR TITLE
agent: support max_tokens, temperature and memory_size options

### PR DESCRIPTION
**Pull request by bdqfork**
## Summary

Add max_tokens, temperature and memory_size options to agent. I did not add this as predict params because it required user always feed to run agent. I think it should be a property of agent more properly, so added in agent creation api.

Fixes

Depends on

## Test plan

-

## Checklist

- [x] My code follows the code style of this project and passes `make format`.
- [x] My change requires a change to the documentation.
- [ ] I have [updated the documentation](https://docs.superagent.sh/) accordingly.
- [ ] My change has adequate unit test coverage.

### CODE REVIEW

**Comment by bdqfork** [comment]
for #72

File: app/api/agents.py [hunk 1]
```diff
@@ -168,7 +171,7 @@ def conversation_run_thread(input: dict) -> None:
                 agent_base = get_agent_base()
                 agent_strategy = AgentFactory.create_agent(agent_base)
                 agent_executor = agent_strategy.get_agent()
-                result = agent_executor(agent_base.process_payload(payload=input))
+                result = agent_executor(input)
```
Comment (homanp) [inline app/api/agents.py:174]: @bdqfork Why is this changed? This will break invocation. Please revert to using `agent_base.process_payload`

**Review by homanp — CHANGES_REQUESTED** [review]
Thanks for this submission @bdqfork!

I have some comments in the PR and a general comment:

The `memorySize`, `maxTokens` and `temperature` should be passed in the existing `llm` dictionary which handles the other llm specific settings such as `model`. 
This way of implementing also means that we don't need to add any fields to the DB and you changes there should not be necessary.

File: app/api/agents.py (hunk 1 above)
Comment (bdqfork) [reply app/api/agents.py:174]: I'm sorry. I forgot to explain. When run agent in ui, I noticed `agent_base.process_payload` will encode `input` to json string, then feed into prompt. For example,  a prompt template like below:
```
User: {input}
```

**Comment by bdqfork** [comment]
> Thanks for this submission @bdqfork!
> 
> I have some comments in the PR and a general comment:
> 
> The `memorySize`, `maxTokens` and `temperature` should be passed in the existing `llm` dictionary which handles the other llm specific settings such as `model`. This way of implementing also means that we don't need to add any fields to the DB and you changes there should not be necessary.

Oh, it's indeed a better solution. Thank you for reminding me.

**Comment by homanp** [comment]
> > Thanks for this submission @bdqfork!
> 
> > 
> 
> > I have some comments in the PR and a general comment:
> 
> > 
> 
> > The `memorySize`, `maxTokens` and `temperature` should be passed in the existing `llm` dictionary which handles the other llm specific settings such as `model`. This way of implementing also means that we don't need to add any fields to the DB and you changes there should not be necessary.
> 
> 
> 
> Oh, it's indeed a better solution. Thank you for reminding me.

Ping when you are ready!

**Comment by bdqfork** [comment]
Already updated! @homanp

**Comment by homanp** [comment]
> Already updated! @homanp 

Thank you!

**Comment by homanp** [comment]
> Already updated! @homanp 

Please revert to using procees_payload as well!

File: app/api/agents.py (hunk 1 above)
Comment (bdqfork) [reply app/api/agents.py]: @homanp I reverted, but I think the input params may be wrong in agent run . The final input to prompt is a json string.

Comment (homanp) [reply app/api/agents.py]: Yeah, should it should be

**Review by homanp — APPROVED** [review]

**Comment by bdqfork** [comment]
Is there anything else to update?

**Comment by homanp** [comment]
> Is there anything else to update?

No thank you! Just testing and then deploying!

**Comment by homanp** [comment]
@bdqfork did you test this on your end?

**Comment by bdqfork** [comment]
I did, is something wrong?发自我的 iPhone在 2023年7月18日，20:37，Ismail Pelaseyed ***@***.***> 写道：﻿
@bdqfork did you test this on your end?

**Comment by homanp** [comment]
@bdqfork did you test the memory changes and new prompt template setup as well?! I didn’t so just want to make sure it will work.

**Comment by homanp** [comment]
> > 
> 
> I did test in my end. However, we do not have a complete regression testing process. If you are also uncertain, I suggest delaying the merge and waiting for the complete testing action to be filled and passed.
> 
> 

Yes ok, I will run a full test and merge after that!

File: app/lib/agents/base.py [hunk 2]
```diff
@@ -111,6 +118,21 @@ def _get_api_key(self) -> str:
...
+                    template=self.prompt.template
+                    if self.prompt
+                    else DEFAULT_CHAT_PROMPT.template
+                )
+                return ChatPromptTemplate(
+                    input_variables=self.prompt.input_variables
+                    if self.prompt
+                    else DEFAULT_CHAT_PROMPT.input_variables,
+                    messages=[system_message_prompt],
+                )
```
[the remark this answers is not in the document]
Comment (bdqfork) [reply app/lib/agents/base.py:132]: In Langchain, the chat model uses different prompt construction methods to differentiate messages from different roles. Please refer to [https://python.langchain.com/docs/modules/model_io/models/chat/how_to/prompts ↗](https://python.langchain.com/docs/modules/model_io/models/chat/how_to/prompts) for more information.

My apologies for the trouble caused. I will submit separate feature requests in the future. Or should I roll back this portion of the code now?

[the remark this answers is not in the document]
Comment (bdqfork) [reply app/lib/agents/base.py:132]: @homanp

[the remark this answers is not in the document]
Comment (homanp) [reply app/lib/agents/base.py:132]: If you could roll back and add this is a seperate PR that would make testing easier.

[the remark this answers is not in the document]
Comment (homanp) [reply app/lib/agents/base.py:132]: If you could roll back and add this is a seperate PR that would make testing easier.

**Comment by bdqfork** [comment]
I will close this pull request. There is some irrelevant code logic included in it, and I will remove it and resubmit a new pull request.

**Comment by homanp** [comment]
> I will close this pull request. There is some irrelevant code logic included in it, and I will remove it and resubmit a new pull request.

Appreciate the effort @bdqfork

**Pull request closed without merging**